### DISCUSSION TITLE
Bitstream for GOWIN FPGA

### DISCUSSION THREAD

**Discussion by dev-hack95** [JulianKemmerer/PipelineC#211] [Q&A] [answered]
Does pipelineC generate bitstream for GOWIN FPGA

**Comment by dev-hack95** [comment]
Its not generating pnr

**Reply by JulianKemmerer** [reply]
It does not generate bitstreams.

I will run PNR if configured to do so
https://github.com/JulianKemmerer/PipelineC/blob/423306ef8c5979ef010a07c6863c77e658e0982b/src/GOWIN.py#L19

but it has been a while since folks have run that tool, I forget if one mode vs the other was fully working

still experimental

**Reply by dev-hack95** [reply]
https://github.com/YosysHQ/apicula.git

Can't we use this should I write code to generate bitstream it's experimental though

**Reply by JulianKemmerer** [reply]
Absolutely hack away doing whatever you want - love hearing ideas. Come over to the Discord to chat if you like :nerd_face: 

But the starting limitation to producing a final bitstream for a board is that pipelinec does not know two critical things for making the final bitstream:
1) the mapping of HDL signals to pins (constraints file)
2) how to instantiate and configure PLLs and such that generate clocks the user needs
(3) sorta like 2^ more generally, if your design needs to instantiate anything outside of pipelinec for whatever reason, you typically do that by hand as you build .bit, some things involving certain FPGA macros/buffers/etc sometimes need this)

**Comment by JulianKemmerer** [comment, marked as answer]
PypelineC does support BitStream gen for a few tools now. See docs.
